tensorflow/java/src/gen/cc/java_defs.h

```diff
@@ -0,0 +1,325 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_JAVA_SRC_GEN_CC_JAVA_DEFS_H_
+#define TENSORFLOW_JAVA_SRC_GEN_CC_JAVA_DEFS_H_
+
+#include <string>
+#include <vector>
+#include <set>
+#include <deque>
+
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/platform/env.h"
+#include "tensorflow/core/lib/core/stringpiece.h"
+
+namespace tensorflow {
+namespace java {
+
+/// \brief An enumeration of different modifiers commonly used in Java
+enum JavaModifier {
+  PUBLIC    = (1 << 0),
+  PROTECTED = (1 << 1),
+  PRIVATE   = (1 << 2),
+  STATIC    = (1 << 3),
+  FINAL     = (1 << 4),
+};
+
+/// \brief A definition of a Java documentation block
+///
+/// Any vector of parameters (@param) that should be included in this block
+/// can be provided separately (e.g. a vector of documented variables, see
+/// JavaVariable).
+class JavaDoc {
+ public:
+  const string& brief() const { return brief_; }
+  JavaDoc& brief(const string& brief) { brief_ = brief; return *this; }
+  const string& description() const { return description_; }
+  JavaDoc& description(const string& txt) { description_ = txt; return *this; }
+  const string& value() const { return value_; }
+  JavaDoc& value(const string& value) { value_ = value; return *this; }
+  bool empty() const {
+    return brief().empty() && description().empty() && value().empty();
+  }
+
+ private:
+  string brief_;
+  string description_;
+  string value_;
+};
+
+class JavaAnnot;
+
+/// \brief A definition of any kind of Java type (classes, interfaces...)
+///
+/// Note that most of the data fields of this class are only useful in specific
+/// contexts and are not required in many cases. For example, annotations and
+/// supertypes are only useful when declaring a type.
+class JavaType {
+ public:
+  enum Kind {
+    PRIMITIVE, CLASS, INTERFACE, GENERIC, ANNOTATION, NONE
```

Comment:
It is only used for empty objects... I tried to focus on simplicity and readability so I tried to avoid dealing with dynamic memory allocation (this code is only run at compile time after all). But looking at it now, the only remaining case of a 'empty' ```JavaType``` is for the return value of a ```JavaMethod```, which should be absent for a constructor. I can probably play a bit around to remove that last case (and the `NONE` value).